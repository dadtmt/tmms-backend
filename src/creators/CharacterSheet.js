import R from 'ramda'
import React, { PropTypes } from 'react'
import { graphql } from 'react-apollo'
import { ListGroup, ListGroupItem } from 'react-bootstrap'
import {
  Field,
  FieldArray,
  propTypes as formPropTypes,
  reduxForm
} from 'redux-form'
import { branch } from 'recompose'

import { pageEditorId } from '../config'
import ButtonGroup from '../forms/components/ButtonGroup'
import CheckboxGroup from '../forms/components/CheckboxGroup'
import FieldGroup from '../forms/components/FieldGroup'
import ImageGroup from '../forms/components/ImageGroup'
import {
  CREATE_SHEET_MUTATION,
  UPDATE_SHEET_MUTATION
} from '../graphql/mutations'

const withCreateMutation = graphql(
  CREATE_SHEET_MUTATION,
  {
    props: ({ mutate }) => ({
      onSubmit: values => mutate({
        variables: {
          newSheet: {
            ...values,
            editorId: pageEditorId
          }
        }
      })
    })
  }
)

const withUpdateMutation = graphql(
  UPDATE_SHEET_MUTATION,
  {
    props: ({ mutate }) => ({
      onSubmit: values => mutate({
        variables: {
          sheet: R.omit('__typename')(values)
        }
      })
    })
  }
)

const renderCharacs = ({ fields }) => <div>
  <ButtonGroup
    glyph='plus'
    onClick={() => fields.push({})}
  />
  <ListGroup>
    {fields.map((charac, index) =>
      <ListGroupItem key={index} >
        <ButtonGroup
          glyph='trash'
          onClick={() => fields.remove(index)}
          title='remove'
        />
        <Field
          component={FieldGroup}
          controlId={`${charac}.name`}
          label='Characteristic:'
          name={`${charac}.name`}
          type='text'
        />
        <Field
          component={FieldGroup}
          controlId={`${charac}.value`}
          label='Value:'
          name={`${charac}.value`}
          type='number'
        />
      </ListGroupItem>
    )}
  </ListGroup>
</div>

renderCharacs.propTypes = {
  fields: PropTypes.object.isRequired
}

const renderGear = ({ fields }) => <div>
  <ButtonGroup
    glyph='plus'
    onClick={() => fields.push({})}
  />
  <ListGroup>
    {fields.map((item, index) =>
      <ListGroupItem key={index} >
        <ButtonGroup
          glyph='trash'
          onClick={() => fields.remove(index)}
          title='remove'
        />
        <Field
          component={FieldGroup}
          controlId={`${item}.name`}
          label='Name:'
          name={`${item}.name`}
          type='text'
        />
        <Field
          component={FieldGroup}
          controlId={`${item}.description`}
          label='Description:'
          name={`${item}.description`}
          type='text'
        />
        <Field
          component={ImageGroup}
          controlId={`${item}.image`}
          label='Image:'
          name={`${item}.image`}
        />
        <Field
          component={FieldGroup}
          controlId={`${item}.count`}
          label='Count:'
          name={`${item}.count`}
          type='number'
        />
      </ListGroupItem>
    )}
  </ListGroup>
</div>

renderGear.propTypes = {
  fields: PropTypes.object.isRequired
}

const CharacterSheet = ({ handleSubmit }) =>
<form onSubmit={handleSubmit}>
  <Field
    component={CheckboxGroup}
    label='player sheet'
    name='isPlayer'
  />
  <Field
    component={FieldGroup}
    controlId='name'
    label='Character name:'
    name='name'
    type='text'
  />
  <Field
    component={FieldGroup}
    controlId='description'
    label='description:'
    name='description'
    type='text'
  />
  <Field
    component={ImageGroup}
    controlId='image'
    label='image:'
    name='image'
  />
  <FieldArray name='characs' component={renderCharacs} />
  <FieldArray name='gear' component={renderGear} />
  <ButtonGroup glyph='save' title='save' type='submit' />
</form>

CharacterSheet.propTypes = {
  ...formPropTypes
}

const CharacterSheetForm = reduxForm({
  enableReinitialize: true,
  form: 'content'
})(CharacterSheet)

export default branch(
  R.pipe(
    R.path(['initialValues', 'id']),
    R.isNil
  ),
  withCreateMutation,
  withUpdateMutation
)(CharacterSheetForm)
