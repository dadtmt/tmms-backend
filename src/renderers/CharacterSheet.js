import React, { PropTypes } from 'react'
import { Label, Image, Panel, Table, Well } from 'react-bootstrap'

const CharacterSheet = ({
  characs,
  description,
  gear,
  image,
  isPlayer,
  name
}) => <Panel header={name}>
  {isPlayer && <Label>Player sheet</Label>}
  <Well>{description}</Well>
  {image && <Image src={image} rounded />}
  <Panel header='Characteristics'>
    <Table>
      <tbody>
        {characs.map(charac => <tr key={charac.name}>
          <td>{charac.name}:</td>
          <td>{charac.value}</td>
        </tr>)}
      </tbody>
    </Table>
  </Panel>
  <Panel header='Gear'>
    <Table>
      <tbody>
        {gear.map(item => <tr key={item.name}>
          <td>{item.name}</td>
          <td>{item.description}</td>
          <td>{item.image && <Image src={item.image} rounded width='50' />}</td>
          <td>{item.count}</td>
        </tr>)}
      </tbody>
    </Table>
  </Panel>
</Panel>

CharacterSheet.propTypes = {
  characs: PropTypes.array.isRequired,
  description: PropTypes.string.isRequired,
  gear: PropTypes.array.isRequired,
  image: PropTypes.string.isRequired,
  isPlayer: PropTypes.bool.isRequired,
  name: PropTypes.string.isRequired
}

export default CharacterSheet
