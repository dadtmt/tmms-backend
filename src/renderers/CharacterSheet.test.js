import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'

import CharacterSheet from './CharacterSheet'

/* eslint-disable */
const props = {
  'id': 'U2hlZXRzOjIz',
  'characs': [
    {
      'name': 'Force',
      'value': '17'
    },
    {
      'name': 'Dex',
      'value': '18'
    }
  ],
  'description': 'Rodeur semi orc',
  'gear': [
    {
      'name': 'Arc Feu',
      'description': '1D8 + 1D6 feu',
      'image': 'https://s-media-cache-ak0.pinimg.com/236x/3b/76/10/3b76104f3387a68e7210c438de6a23c7.jpg',
      'count': '1'
    },
    {
      'name': 'Fleches',
      'count': '10'
    }
  ],
  'image': 'https://s-media-cache-ak0.pinimg.com/736x/fe/a2/b2/fea2b2e27dc64aed4228bb3697018436.jpg',
  'isPlayer': true,
  'name': 'Marom',
  '__typename': 'Sheets'
}

/* eslint-enable */

it('renders without crashing', () => {
  const div = document.createElement('div')

  ReactDOM.render(
    <CharacterSheet {...props} />,
    div
  )
})

it('renders correctly', () => {
  expect(
    renderer.create(<CharacterSheet {...props} />).toJSON()
  ).toMatchSnapshot()
})
