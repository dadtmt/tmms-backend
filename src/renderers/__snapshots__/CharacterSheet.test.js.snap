exports[`test renders correctly 1`] = `
<div
  className="panel panel-default"
  id={undefined}>
  <div
    className="panel-heading">
    Marom
  </div>
  <div
    className="panel-body">
    <span
      className="label label-default">
      Player sheet
    </span>
    <div
      className="well">
      Rodeur semi orc
    </div>
    <img
      className="img-rounded"
      src="https://s-media-cache-ak0.pinimg.com/736x/fe/a2/b2/fea2b2e27dc64aed4228bb3697018436.jpg" />
    <table
      className="table">
      <tbody>
        <tr>
          <td>
            Force
            :
          </td>
          <td>
            17
            :
          </td>
        </tr>
        <tr>
          <td>
            Dex
            :
          </td>
          <td>
            18
            :
          </td>
        </tr>
      </tbody>
    </table>
    <table
      className="table">
      <tbody>
        <tr>
          <td>
            Arc Feu
          </td>
          <td>
            1D8 + 1D6 feu
          </td>
          <td>
            <img
              className="img-rounded"
              src="https://s-media-cache-ak0.pinimg.com/236x/3b/76/10/3b76104f3387a68e7210c438de6a23c7.jpg" />
          </td>
          <td>
            1
          </td>
        </tr>
        <tr>
          <td>
            Fleches
          </td>
          <td />
          <td />
          <td>
            10
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
